src/index.js

```diff
@@ -58,45 +58,59 @@ module.exports = class FaviconsWebpackPlugin {
 
     // Hook into the webpack compilation
     // to start the favicon generation
-    tap(compiler, 'make', 'FaviconsWebpackPlugin', (compilation, callback) => {
-      let faviconCompilation;
-      switch (this.getCurrentCompilationMode(compiler) ) {
-        case 'light':
-          faviconCompilation = this.generateFaviconsLight(compiler, compilation);
-          break;
-        case 'webapp':
-        default:
-          faviconCompilation = this.generateFaviconsWebapp(compiler, compilation);
-      }
-      
+    compiler.hooks.make.tapPromise('FaviconsWebpackPlugin', async compilation => {
+      const faviconCompilation = this.generateFavicons(compiler, compilation);
+
       // Hook into the html-webpack-plugin processing and add the html
-      tapHtml(compilation, 'FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
-        faviconCompilation.then((tags) => {
-          if (this.options.inject(htmlPluginData.plugin)) {
-            const idx = (htmlPluginData.html + '</head>').search(/<\/head>/i);
-            htmlPluginData.html = [htmlPluginData.html.slice(0, idx), ...tags, htmlPluginData.html.slice(idx)].join('');
-          }
-          htmlWebpackPluginCallback(null, htmlPluginData);
-        }).catch(htmlWebpackPluginCallback);
-      });
+      const htmlWebpackPlugin = compiler.options.plugins
+        .map(({ constructor }) => constructor)
+        .find(({ name }) => name === 'HtmlWebpackPlugin');
+
+      if (htmlWebpackPlugin) {
+        htmlWebpackPlugin.getHooks(compilation).alterAssetTags.tapAsync('FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
+          faviconCompilation.then((tags) => {
 
-      // Save the promise and execute the callback immediately to not block 
-      // the webpack build see the `afterCompile` FaviconsWebpackPlugin hook 
+            if (this.options.inject(htmlPluginData.plugin)) {
+              htmlPluginData.plugin.options.inject = true;
+              [].push.apply(
```

Comment:
What about using the spread operator for better readability?  
It is supported for node 8

```js
htmlPluginData.assetTags.meta.push(...tags.map
```

Reply:
Sure.